src/types/quota.rs

```diff
@@ -0,0 +1,160 @@
+use crate::error::{Error, ParseError};
+use crate::parse::{parse_many_into, try_handle_unilateral, MapOrNot};
+use crate::types::UnsolicitedResponse;
+use imap_proto::Response;
+use ouroboros::self_referencing;
+use std::fmt::{Display, Formatter};
+use std::sync::mpsc;
+
+pub struct QuotaLimit {
+    // The resource type
+    pub name: QuotaLimitName,
+    // The amount for that resource
+    pub amount: u64,
+}
+
+impl Display for QuotaLimit {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{} {}", self.name, self.amount)
+    }
+}
+
+/// https://datatracker.ietf.org/doc/html/rfc2087#section-3
+pub enum QuotaLimitName {
+    /// Sum of messages' RFC822.SIZE, in units of 1024 octets
+    Storage,
+    /// Number of messages
+    Message,
+    /// Any other string (for future RFCs)
+    Atom(String),
+}
+
+impl Display for QuotaLimitName {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            QuotaLimitName::Storage => write!(f, "STORAGE"),
+            QuotaLimitName::Message => write!(f, "MESSAGE"),
+            QuotaLimitName::Atom(s) => write!(f, "{}", s),
+        }
+    }
+}
+
+#[self_referencing]
+pub struct Quota {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) quota: imap_proto::Quota<'this>,
+}
+
+impl Quota {
+    pub fn parse(
+        owned: Vec<u8>,
+        unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
+    ) -> Result<Self, Error> {
+        QuotaTryBuilder {
+            data: owned,
+            quota_builder: |input| {
+                let mut lines: &[u8] = input;
+
+                // There should only be one single QUOTA response
+                while !lines.is_empty() {
+                    match imap_proto::parser::parse_response(lines) {
+                        Ok((_rest, Response::Quota(a))) => {
+                            return Ok(a);
```

Comment:
As we discussed in the other PR, this probably shouldn't be early-returning.

Reply:
yeah.. already have this reworked in my next push